src/tools/kdb/command.c

```diff
@@ -0,0 +1,112 @@
+/**
+ * @file
+ *
+ * @brief Implementation of things used everywhere in the kdb tool
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <command.h>
+
+#include <kdberrors.h>
+#include <stdio.h>
+#include <string.h>
+
+const char * expandKeyName (const char * name, bool * resolved)
+{
+	*resolved = false;
+	const char * ret = NULL;
+	if (name == NULL)
+	{
+		return ret;
+	}
+
+	KeySet * bookmarks = ksNew (0, KS_END);
+	Key * bookmarkBase = keyNew (CLI_BASE_KEY "/bookmarks", KEY_END);
+	KDB * handle = kdbOpen (NULL, bookmarkBase);
+	kdbGet (handle, bookmarks, bookmarkBase);
```

Comment:
I don't know how often this function is actually called. But this hidden `kdbGet` seems problematic. Especially since it always uses the same parent key. It would better if `kdb` loaded the bookmarks once at the start and then simply used the existing `KeySet`.

Also since `bookmarks` is only used in one of the branches below, please move the whole KDB stuff to that branch.

Or just do

```c
	if (name[0] != '+')
	{
		return elektraStrDup (name);
	}
```

before you start the bookmarks stuff.

Reply:
Moved it into branch. This is called in a few commands, and I think in one or two twice.